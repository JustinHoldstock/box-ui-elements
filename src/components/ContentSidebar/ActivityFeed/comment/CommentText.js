/**
 * @flow
 * @file Comment Text component used by Comment component
 */

import React, { Component, ReactNode } from 'react';

import LoadingIndicator from 'box-react-ui/lib/components/loading-indicator';

import formatTaggedMessage from '../utils/formatTaggedMessage';
import ShowOriginalButton from './ShowOriginalButton';
import TranslateButton from './TranslateButton';

type Props = {
    id: string,
    taggedMessage: string,
    translatedTaggedMessage: string,
    translationEnabled: boolean,
    onTranslate: Function,
    translationFailed: boolean
};

class CommentText extends Component<Props> {
    static defaultProps = {
        translationEnabled: false
    };

    state = {
        isLoading: false,
        isTranslation: false
    };

    componentWillReceiveProps(nextProps: Props): void {
        const { translatedTaggedMessage, translationFailed } = nextProps;
        if (translatedTaggedMessage || translationFailed) {
            this.setState({ isLoading: false });
        }
    }

    getButton(isTranslation: boolean): ReactNode {
        let button = null;
        if (isTranslation) {
            button = <ShowOriginalButton handleShowOriginal={this.handleShowOriginal} />;
        } else {
            button = <TranslateButton handleTranslate={this.handleTranslate} />;
        }
        return button;
    }

    handleTranslate = (event: React.SyntheticEvent): void => {
        const { id, taggedMessage, onTranslate, translatedTaggedMessage } = this.props;
        if (!translatedTaggedMessage) {
            this.setState({ isLoading: true });
            onTranslate({ id, taggedMessage });
        }
        this.setState({ isTranslation: true });
        event.preventDefault();
    };

    handleShowOriginal = (event: React.SyntheticEvent): void => {
        this.setState({ isTranslation: false });
        event.preventDefault();
    };

    render(): ReactNode {
        const { id, taggedMessage, translatedTaggedMessage, translationEnabled } = this.props;
        const { isLoading, isTranslation } = this.state;
        const commentToDisplay =
            translationEnabled && isTranslation && translatedTaggedMessage ? translatedTaggedMessage : taggedMessage;
        return isLoading ? (
            <div className='bcs-comment-text-loading'>
                <LoadingIndicator size='small' />
            </div>
        ) : (
            <div className='bcs-comment-text'>
                {formatTaggedMessage(commentToDisplay, id, false)}
                {translationEnabled ? this.getButton(isTranslation) : null}
            </div>
        );
    }
}

export default CommentText;
