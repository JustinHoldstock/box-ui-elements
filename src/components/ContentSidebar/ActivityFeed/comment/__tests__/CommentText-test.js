import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';

import CommentText from '../CommentText';

const sandbox = sinon.sandbox.create();

describe('features/activity-feed/comment/CommentText', () => {
    afterEach(() => {
        sandbox.verifyAndRestore();
    });

    test('should properly format tagged comment', () => {
        const commentText = {
            taggedMessage: 'How u doing @[2030326577:Young Jeezy]?'
        };

        const wrapper = shallow(<CommentText id='123' {...commentText} />);

        expect(wrapper.find('.box-ui-comment-text').prop('dangerouslySetInnerHTML')).toBeFalsy();
        expect(wrapper.find('.box-ui-comment-text').find('Mention').length).toEqual(1);
    });

    test('should properly handle unicode variants of @ in tagged comments', () => {
        const commentText = {
            taggedMessage: 'Hi ﹫[123:Half] ＠[222:Full] @[432:Latin]'
        };

        const wrapper = shallow(<CommentText id='123' {...commentText} />);

        expect(wrapper.find('.box-ui-comment-text').prop('dangerouslySetInnerHTML')).toBeFalsy();
        expect(wrapper.find('.box-ui-comment-text').find('Mention').length).toEqual(3);
    });

    test('should not show translate button by default, translation should be disabled', () => {
        const commentText = { taggedMessage: 'test' };

        const wrapper = mount(<CommentText id='123' {...commentText} />);

        expect(wrapper.find('.box-ui-comment-text').text()).toEqual(commentText.taggedMessage);
        expect(wrapper.find('PlainButton.box-ui-comment-translate').length).toEqual(0);
        expect(wrapper.prop('translationEnabled')).toBe(false);
        expect(wrapper.state('isLoading')).toBe(false);
    });

    test('should show translate button when translation is enabled', () => {
        const translations = { translationEnabled: true };
        const commentText = { taggedMessage: 'test' };

        const wrapper = mount(<CommentText id='123' {...commentText} {...translations} />);

        expect(wrapper.find('PlainButton.box-ui-comment-translate').length).toEqual(1);
        expect(wrapper.state('isTranslation')).toBe(false);
        expect(wrapper.state('isLoading')).toBe(false);
    });

    test('should show original button when translation is enabled and already showing translated comment', () => {
        const translations = { translationEnabled: true };
        const commentText = {
            taggedMessage: 'test',
            translatedTaggedMessage: 'translated'
        };

        const wrapper = mount(<CommentText id='123' {...commentText} {...translations} />);
        wrapper.setState({ isTranslation: true });

        expect(wrapper.find('PlainButton.box-ui-comment-translate').length).toEqual(1);
        expect(wrapper.state('isLoading')).toBe(false);
    });

    test('should show loading indicator when state is isLoading', () => {
        const translations = { translationEnabled: true };
        const commentText = {
            taggedMessage: 'test',
            translatedTaggedMessage: 'translated'
        };

        const wrapper = shallow(<CommentText id='123' {...commentText} {...translations} />);
        wrapper.setState({ isTranslation: false, isLoading: true });

        expect(wrapper.find('LoadingIndicator').length).toEqual(1);
    });

    test('should call onTranslate when translate button is clicked', () => {
        const onTranslateSpy = sandbox.spy();
        const translations = {
            translationEnabled: true,
            onTranslate: onTranslateSpy
        };
        const commentText = { taggedMessage: 'test' };

        const wrapper = mount(<CommentText id='123' {...commentText} {...translations} />);

        const translateBtn = wrapper.find('PlainButton.box-ui-comment-translate');
        translateBtn.simulate('click');

        expect(onTranslateSpy.calledOnce).toBe(true);
        expect(wrapper.find('.box-ui-comment-text-loading').length).toEqual(1);
        expect(wrapper.state('isTranslation')).toBe(true);
        expect(wrapper.state('isLoading')).toBe(true);
    });

    test('should not call onTranslate when translate button is clicked and translated comment exists', () => {
        const onTranslateSpy = sandbox.spy();
        const translations = {
            translationEnabled: true,
            onTranslate: onTranslateSpy
        };
        const commentText = {
            taggedMessage: 'test',
            translatedTaggedMessage: 'translated'
        };

        const wrapper = mount(<CommentText id='123' {...commentText} {...translations} />);
        wrapper.setState({ isTranslation: false });

        const translateBtn = wrapper.find('PlainButton.box-ui-comment-translate');
        translateBtn.simulate('click');

        expect(onTranslateSpy.callCount).toEqual(0);
        expect(wrapper.find('PlainButton.box-ui-comment-translate').length).toEqual(1);
        expect(wrapper.state('isTranslation')).toBe(true);
        expect(wrapper.state('isLoading')).toBe(false);
    });

    test('should show comment when show original button is clicked', () => {
        const onTranslateSpy = sandbox.spy();
        const translations = {
            translationEnabled: true,
            onTranslate: onTranslateSpy
        };
        const commentText = {
            taggedMessage: 'test',
            translatedTaggedMessage: 'translated'
        };

        const wrapper = mount(<CommentText id='123' {...commentText} {...translations} />);
        wrapper.setState({ isTranslation: true });

        const showOriginalBtn = wrapper.find('PlainButton.box-ui-comment-translate');
        showOriginalBtn.simulate('click');

        expect(onTranslateSpy.callCount).toEqual(0);
        expect(wrapper.find('PlainButton.box-ui-comment-translate').length).toEqual(1);
        expect(wrapper.state('isTranslation')).toBe(false);
        expect(wrapper.state('isLoading')).toBe(false);
    });
});
