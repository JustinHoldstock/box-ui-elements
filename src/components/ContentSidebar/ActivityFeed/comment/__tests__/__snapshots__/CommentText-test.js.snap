// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/activity-feed/comment/CommentText should properly format tagged comment 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CommentText
    id="123"
    taggedMessage="How u doing @[2030326577:Young Jeezy]?"
    translationEnabled={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          "How u doing ",
          <Mention
            id={2030326577}
          >
            @Young Jeezy
          </Mention>,
          "?",
        ],
        null,
      ],
      "className": "bcs-comment-text",
    },
    "ref": null,
    "rendered": Array [
      "How u doing ",
      Object {
        "instance": null,
        "key": "1-123",
        "nodeType": "function",
        "props": Object {
          "children": "@Young Jeezy",
          "id": 2030326577,
        },
        "ref": null,
        "rendered": "@Young Jeezy",
        "type": [Function],
      },
      "?",
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            "How u doing ",
            <Mention
              id={2030326577}
            >
              @Young Jeezy
            </Mention>,
            "?",
          ],
          null,
        ],
        "className": "bcs-comment-text",
      },
      "ref": null,
      "rendered": Array [
        "How u doing ",
        Object {
          "instance": null,
          "key": "1-123",
          "nodeType": "function",
          "props": Object {
            "children": "@Young Jeezy",
            "id": 2030326577,
          },
          "ref": null,
          "rendered": "@Young Jeezy",
          "type": [Function],
        },
        "?",
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`features/activity-feed/comment/CommentText should properly handle unicode variants of @ in tagged comments 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CommentText
    id="123"
    taggedMessage="Hi ﹫[123:Half] ＠[222:Full] @[432:Latin]"
    translationEnabled={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          "Hi ",
          <Mention
            id={123}
          >
            ﹫Half
          </Mention>,
          " ",
          <Mention
            id={222}
          >
            ＠Full
          </Mention>,
          " ",
          <Mention
            id={432}
          >
            @Latin
          </Mention>,
          "",
        ],
        null,
      ],
      "className": "bcs-comment-text",
    },
    "ref": null,
    "rendered": Array [
      "Hi ",
      Object {
        "instance": null,
        "key": "1-123",
        "nodeType": "function",
        "props": Object {
          "children": "﹫Half",
          "id": 123,
        },
        "ref": null,
        "rendered": "﹫Half",
        "type": [Function],
      },
      " ",
      Object {
        "instance": null,
        "key": "3-123",
        "nodeType": "function",
        "props": Object {
          "children": "＠Full",
          "id": 222,
        },
        "ref": null,
        "rendered": "＠Full",
        "type": [Function],
      },
      " ",
      Object {
        "instance": null,
        "key": "5-123",
        "nodeType": "function",
        "props": Object {
          "children": "@Latin",
          "id": 432,
        },
        "ref": null,
        "rendered": "@Latin",
        "type": [Function],
      },
      "",
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            "Hi ",
            <Mention
              id={123}
            >
              ﹫Half
            </Mention>,
            " ",
            <Mention
              id={222}
            >
              ＠Full
            </Mention>,
            " ",
            <Mention
              id={432}
            >
              @Latin
            </Mention>,
            "",
          ],
          null,
        ],
        "className": "bcs-comment-text",
      },
      "ref": null,
      "rendered": Array [
        "Hi ",
        Object {
          "instance": null,
          "key": "1-123",
          "nodeType": "function",
          "props": Object {
            "children": "﹫Half",
            "id": 123,
          },
          "ref": null,
          "rendered": "﹫Half",
          "type": [Function],
        },
        " ",
        Object {
          "instance": null,
          "key": "3-123",
          "nodeType": "function",
          "props": Object {
            "children": "＠Full",
            "id": 222,
          },
          "ref": null,
          "rendered": "＠Full",
          "type": [Function],
        },
        " ",
        Object {
          "instance": null,
          "key": "5-123",
          "nodeType": "function",
          "props": Object {
            "children": "@Latin",
            "id": 432,
          },
          "ref": null,
          "rendered": "@Latin",
          "type": [Function],
        },
        "",
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
